### PR TITLE
Squash network inbound queues

### PR DESCRIPTION
**Pull request by brianolson**
## Description
Remove several queues of indirection, move received messages directly from the reader thread of a peer to application code.

add inbound queue delay metric `aptos_network_inbound_queue_time`

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [X] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [X] Validator Node
- [X] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
Local cluster tests. unit tests. forge cluster tests.

## Key Areas to Review
Some unit tests in mempool were lost due to relying on layers of the network that are being phased out.

Some parts of the inbound path may still be there and unused, further code cleanup is possible.

## Checklist
- [X] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I identified and added all stakeholders and component owners affected by this change as reviewers
- [X] I tested both happy and unhappy path of the functionality
- [X] I have made corresponding changes to the documentation

### CODE REVIEW

File: network/framework/src/protocols/health_checker/test.rs [hunk 1]
```diff
@@ -132,21 +135,30 @@ impl TestHarness {
...
-            .into();
+        let data = bcs::to_bytes(&HealthCheckerMsg::Ping(Ping(ping))).unwrap();
+        // .into();
         let (res_tx, res_rx) = oneshot::channel();
-        let inbound_rpc_req = InboundRpcRequest {
-            protocol_id,
-            data,
-            res_tx,
-        };
+        // let inbound_rpc_req = InboundRpcRequest {
```
Comment (JoshLind) [inline network/framework/src/protocols/health_checker/test.rs:141]: ?

File: network/framework/src/peer_manager/types.rs [hunk 2]
```diff
@@ -26,6 +26,7 @@ pub enum PeerManagerRequest {
 }
 
 /// Notifications sent by PeerManager to upstream actors.
+/// TODO: PeerManagerNotification now only exists in test code and should be deleted; probably use `ReceivedMessage`
```
Comment (JoshLind) [inline network/framework/src/peer_manager/types.rs:29]: Nice 😄

**Review by JoshLind — COMMENTED** [review]
Quick pass. Will return to look at the internals 😄

File: mempool/src/tests/node.rs [hunk 3]
```diff
@@ -75,9 +36,9 @@ impl NodeId {
 /// Validators, ValidatorFullNodes, and FullNodes
 #[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialOrd, PartialEq)]
 pub enum NodeType {
-    Validator,
-    ValidatorFullNode,
-    FullNode,
+    // Validator,
```
[the remark this answers is not in the document]
Comment (brianolson) [reply mempool/src/tests/node.rs:39]: lint complained that none of these are used, but it also felt weird to just delete them and leave an empty enum

File: network/framework/src/counters.rs [hunk 4]
```diff
@@ -638,3 +638,19 @@ pub static OP_MEASURE: Lazy<HistogramVec> = Lazy::new(|| {
...
+        &["protocol_id"],
+        exponential_buckets(/*start=*/ 1e-6, /*factor=*/ 2.0, /*count=*/ 20).unwrap(),
+    )
+    .unwrap()
+});
+
+pub fn inbound_queue_delay_observe(protocol_id: ProtocolId, micros: u64) {
+    INBOUND_QUEUE_DELAY
+        .with_label_values(&[protocol_id.as_str()])
+        .observe((micros as f64) / 1000000.0)
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/counters.rs]: okay, that's more consistent, done

File: consensus/src/network_tests.rs [hunk 5]
```diff
@@ -496,7 +507,7 @@ mod tests {
             storage::PeersAndMetadata,
         },
         protocols::{
-            direct_send::Message,
+            // direct_send::Message,
```
[the remark this answers is not in the document]
Comment (brianolson) [reply consensus/src/network_tests.rs]: cleaned up

File: network/framework/src/protocols/network/mod.rs [hunk 6]
```diff
@@ -133,13 +132,66 @@ impl NetworkApplicationConfig {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct ReceivedMessage {
+    pub message: NetworkMessage,
+    pub sender: PeerNetworkId,
+
+    // unix microseconds
+    pub rx_at: u64,
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/protocols/network/mod.rs]: `receive_timestamp_micros`

File: network/framework/src/protocols/health_checker/test.rs [hunk 7]
```diff
@@ -132,21 +135,25 @@ impl TestHarness {
         ping: u32,
     ) -> oneshot::Receiver<Result<Bytes, RpcError>> {
         let protocol_id = ProtocolId::HealthCheckerRpc;
-        let data = bcs::to_bytes(&HealthCheckerMsg::Ping(Ping(ping)))
-            .unwrap()
-            .into();
+        let data = bcs::to_bytes(&HealthCheckerMsg::Ping(Ping(ping))).unwrap();
+        // .into();
```
Comment (JoshLind) [inline network/framework/src/protocols/health_checker/test.rs:139]: ?

**Review by JoshLind — COMMENTED** [review]
Looks reasonable to me. 😄 Will wait for @bchocho to comment.

File: mempool/src/tests/node.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JoshLind) [reply mempool/src/tests/node.rs:39]: Aah, I see. In that case, something like `#[allow(unused)]` or `#[allow(dead_code)]` might be better to silence the linter 😄 (At least until we can remove all the dead code)

File: network/framework/src/peer/test.rs [hunk 8]
```diff
@@ -480,24 +551,23 @@ fn peer_recv_rpc_timeout() {
...
 
     let test = async move {
         // Client sends the rpc request.
         client_sink.send(&send_msg).await.unwrap();
 
         // Server receives the rpc request from client.
-        let received = peer_notifs_rx.next().await.unwrap();
-        assert_eq!(received, recv_msg);
+        let received = prot_rx.next().await.unwrap();
+        // assert_eq!(received, recv_msg);
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer/test.rs:561]: replaced with an assert a few lines below that checks that the message content is correct

File: network/framework/src/peer/test.rs [hunk 9]
```diff
@@ -243,41 +253,53 @@ fn peer_send_message() {
...
     });
 
     let client = async move {
+        info!("client start");
         let mut connection = MultiplexMessageSink::new(connection, MAX_FRAME_SIZE);
         for _ in 0..30 {
             // The client should then send the network message.
             connection.send(&send_msg).await.unwrap();
         }
+        info!("client sent");
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer/test.rs:281]: test code logging is pretty cheap? it gets thrown away if the test passes, right?

File: network/framework/src/peer/test.rs [hunk 10]
```diff
@@ -286,10 +308,23 @@ fn peer_recv_message() {
 fn peers_send_message_concurrent() {
     ::aptos_logger::Logger::init_for_testing();
     let rt = Runtime::new().unwrap();
+    let mut upstream_handlers_a = HashMap::new();
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer/test.rs]: squashed into a utility function

File: network/framework/src/protocols/network/mod.rs [hunk 11]
```diff
@@ -216,29 +268,47 @@ impl<TMessage> Stream for NetworkEvents<TMessage> {
...
-        PeerManagerNotification::RecvRpc(peer_id, rpc_req) => {
+    let peer_id = message.sender.peer_id();
+    let ReceivedMessage {
+        message,
+        sender: _sender,
+        receive_timestamp_micros: rx_at,
+        rpc_replier,
+    } = message;
+    let now = std::time::SystemTime::now();
+    let dequeue_at = now
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/protocols/network/mod.rs]: made a `unix_micros()` util

File: mempool/src/tests/node.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (brianolson) [reply mempool/src/tests/node.rs]: Changed to `#[allow(unused)]` and added an extra comment

File: network/framework/src/protocols/health_checker/test.rs (hunk 7 above)
Comment (brianolson) [reply network/framework/src/protocols/health_checker/test.rs]: cleaned up

File: network/framework/src/peer/test.rs [hunk 12]
```diff
@@ -74,8 +83,10 @@ fn build_test_peer(
     let (connection_notifs_tx, connection_notifs_rx) = aptos_channels::new_test(1);
     let (peer_reqs_tx, peer_reqs_rx) =
         aptos_channel::new(QueueStyle::FIFO, NETWORK_CHANNEL_SIZE, None);
-    let (peer_notifs_tx, peer_notifs_rx) =
-        aptos_channel::new(QueueStyle::FIFO, NETWORK_CHANNEL_SIZE, None);
+    // let (peer_notifs_tx, peer_notifs_rx) =
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer/test.rs]: cleaned up dead code

File: network/framework/src/peer/test.rs [hunk 13]
```diff
@@ -538,24 +612,23 @@ fn peer_recv_rpc_cancel() {
...
 
     let test = async move {
         // Client sends the rpc request.
         client_sink.send(&send_msg).await.unwrap();
 
         // Server receives the rpc request from client.
-        let received = peer_notifs_rx.next().await.unwrap();
-        assert_eq!(received, recv_msg);
+        let received = prot_rx.next().await.unwrap();
+        // assert_eq!(received, recv_msg);
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer/test.rs]: changed to an assert below to check the message content

**Comment by brianolson** [comment]
I took a stab at mempool fixing and it wasn't as bad as the full network2 branch with changing both inbound and outbound paths. Merged the fix in here. `aptos-mempool` tests enabled and passing again.

**Review by bchocho — COMMENTED** [review]
`mempool` changes look great! I'll continue looking at the rest of the PR

File: network/framework/src/protocols/network/mod.rs [hunk 14]
```diff
@@ -214,31 +262,52 @@ impl<TMessage> Stream for NetworkEvents<TMessage> {
     }
 }
 
+fn unix_micros() -> u64 {
```
Comment (bchocho) [inline network/framework/src/protocols/network/mod.rs:265]: Any reason not to use time service? https://github.com/aptos-labs/aptos-core/blob/ce056658101865b91af44078941c8ea36458119d/crates/aptos-time-service/src/lib.rs#L155

Comment (brianolson) [reply network/framework/src/protocols/network/mod.rs:265]: the usage sites are pretty awkward to plumb that through to

File: network/framework/src/peer_manager/mod.rs [hunk 15]
```diff
@@ -109,7 +109,8 @@ where
         HashMap<ConnectionId, oneshot::Sender<Result<(), PeerManagerError>>>,
     /// Pin the transport type corresponding to this PeerManager instance
     phantom_transport: PhantomData<TTransport>,
-    /// Maximum concurrent network requests to any peer.
+    /// Maximum concurrent network requests to any peer. // TODO: bring back usage?
```
[the remark this answers is not in the document]
Comment (brianolson) [reply network/framework/src/peer_manager/mod.rs]: removed (and several levels of plumbing through the now unused value)

**Review by bchocho — APPROVED** [review]
LGTM!

File: network/framework/src/counters.rs [hunk 16]
```diff
@@ -26,6 +26,7 @@ pub const RECEIVED_LABEL: &str = "received";
 pub const SENT_LABEL: &str = "sent";
 pub const SUCCEEDED_LABEL: &str = "succeeded";
 pub const FAILED_LABEL: &str = "failed";
+pub const UNKNOWN_LABEL: &str = "unk";
```
Comment (JoshLind) [inline network/framework/src/counters.rs:29]: nit: can we just make this `unknown`? 😄

File: network/framework/src/peer_manager/tests.rs [hunk 17]
```diff
@@ -124,7 +122,7 @@ fn build_test_peer_manager(
         peer_manager,
         peer_manager_request_tx,
         connection_reqs_tx,
-        hello_rx,
+        // hello_rx,
```
Comment (JoshLind) [inline network/framework/src/peer_manager/tests.rs:125]: nit: remove?

File: network/framework/src/peer/mod.rs [hunk 18]
```diff
@@ -461,21 +496,37 @@ where
...
+                                self.network_context,
+                                err
+                            );
+                        }
+                    },
                 }
             },
-            NetworkMessage::RpcResponse(response) => {
+            NetworkMessage::RpcResponse(_) => {
+                // non-reference cast identical to this match case
```
Comment (JoshLind) [inline network/framework/src/peer/mod.rs:526]: What's the reason we can't just match this directly?

File: network/framework/src/protocols/network/mod.rs [hunk 19]
```diff
@@ -214,31 +262,52 @@ impl<TMessage> Stream for NetworkEvents<TMessage> {
...
+    let peer_id = message.sender.peer_id();
+    let ReceivedMessage {
+        message,
+        sender: _sender,
+        receive_timestamp_micros: rx_at,
+        rpc_replier,
+    } = message;
+    let dequeue_at = unix_micros();
+    let dt_micros = dequeue_at - rx_at;
+    let dt_seconds = (dt_micros as f64) / 1000000.0;
```
Comment (JoshLind) [inline network/framework/src/protocols/network/mod.rs:286]: nit: I'd move this to a helper:
```
let dequeue_delta_secs = calculate_dequeue_delta(rx_at);
...
fn calculate_dequeue_delta(rx_at: u64) -> u64 {
    let dequeue_at = unix_micros();
    let dt_micros = dequeue_at - rx_at;
    (dt_micros as f64) / 1000000.0
}
```

**Review by JoshLind — APPROVED** [review]
Thanks @brianolson 😄

File: network/framework/src/peer/mod.rs (hunk 18 above)
Comment (brianolson) [reply network/framework/src/peer/mod.rs:526]: the other branches work better with a match over a reference, and `*response` didn't work.

**Pull request merged**